### PR TITLE
Add a registry of additional config checks to check inter-parameter incompatibilities.

### PR DESCRIPTION
**Pull request by justinxzhao**
## Overview

There are many interdependent parameters in a ludwig configuration, and generating a marshmallow schema for incompatible interdependent parameters can be much more awkward/difficult than a straightforward traversal through a python dictionary.

There are also auxiliary checks scattered throughout the codebase that raise errors upon discovering incompatible parameter values.

This PR attempts to consolidate the lion's share of them in a single registry, and each check gets called as part of `validate_config`.

## Validation API

After this PR, it might be a good idea to fold all validation into the `from_dict()` method, so that there's one API for getting a comprehensive and validated config. 

```python
comprehensive_and_validated_config = ModelConfig.from_dict(user_config)
```

## List of consolidated and new checks

These now exist as part of the `config_check_registry`:
- `check_feature_names_unique`: Checks that all feature names are unique.
- `check_tied_features_are_valid`: Checks that all tied features are valid.
- `check_training_runway`: Checks that checkpoints_per_epoch and steps_per_checkpoint aren't simultaneously defined.
- `check_gbm_horovod_incompatibility`: Checks that GBM model type isn't being used with the horovod backend.
- `check_gbm_single_output_feature`: GBM models only support a single output feature.
- `check_gbm_feature_types`: Checks that all input features for GBMs are of supported types.
- `check_ray_backend_in_memory_preprocessing`: Checks that in memory preprocessing is used with Ray backend.
- `check_sequence_concat_combiner_requirements`: Checks that sequence concat combiner has at least one input feature that's sequential.
- `check_comparator_combiner_requirements`: Checks ComparatorCombiner requirements. All of the feature names for entity_1 and entity_2 are valid features.
- `check_class_balance_preprocessing`: Class balancing is only available for datasets with a single output feature.
- `check_sampling_exclusivity`: Oversample minority and undersample majority are mutually exclusive.
- `check_validation_metric_exists`: Checks that validation fields in config.trainer are valid.
- `check_splitter`: Checks the validity of the splitter configuration.

### CODE REVIEW

File: ludwig/schema/model_config.py
Comment (ksbrar) [inline ludwig/schema/model_config.py:270]: Nit: for brevity, you could create a registry of auxiliary checks

Comment (justinxzhao) [reply ludwig/schema/model_config.py]: Done.

File: ludwig/config_validation/checks.py
Comment (ksbrar) [inline ludwig/config_validation/checks.py:40]: worth moving to `ludwig.constants`?

File: ludwig/config_validation/checks.py
Comment (ksbrar) [inline ludwig/config_validation/checks.py:257]: Hmm, jw - does this cause the schema to be validated twice?

Comment (justinxzhao) [reply ludwig/config_validation/checks.py:257]: Instantiating a model config will run schema validation twice -- once at the beginning of the constructor to check required fields (the rest of the config class code was written assuming that required fields are all present), and once at the end when we iterate through the registry.

File: ludwig/config_validation/checks.py
Comment (justinxzhao) [reply ludwig/config_validation/checks.py:40]: I don't see any immediate use for this being used outside of this file. Happy to do that refactoring in the future when the need arises.

File: ludwig/config_validation/checks.py
Comment (ksbrar) [reply ludwig/config_validation/checks.py:257]: Why do it the second time?

**Comment by tgaddair** [comment]
It might make sense to hold off on some of these changes until #2906 lands, which makes some major changes to config object.

File: ludwig/config_validation/checks.py
Comment (justinxzhao) [reply ludwig/config_validation/checks.py]: Removed the second schema check.

File: ludwig/config_validation/checks.py
Comment (ksbrar) [inline ludwig/config_validation/checks.py:86]: I still wonder if we should consider exposing the auxiliary checks through a separate API so they can be used in the FE as well. Flagging here to not lose track of that thread, but this is great.

**Review by ksbrar — APPROVED** [review]
clean 👌🏽

File: ludwig/config_validation/checks.py
Comment (tgaddair) [inline ludwig/config_validation/checks.py:155]: I would remove this one since it's in the schema already and I would prefer to keep a single source of turth ofr this.

File: ludwig/config_validation/checks.py
Comment (tgaddair) [inline ludwig/config_validation/checks.py:195]: This is fine for now because we don't validate on the backend, but can be removed in the future since we have separate schemas for ECD and GBM.

File: ludwig/config_validation/checks.py
Comment (tgaddair) [inline ludwig/config_validation/checks.py:206]: This is already in jsonschema / marshmallow, so let's remove.

File: ludwig/config_validation/checks.py
Comment (tgaddair) [inline ludwig/config_validation/checks.py:214]: This is also in the schema already. Remove.

File: ludwig/config_validation/checks.py
Comment (ksbrar) [reply ludwig/config_validation/checks.py:155]: @tgaddair Where in the schema? If you are referring to `uniqueItemProperties` remember that that's not an official JSON keyword - merely an extra one that AJV supports. So we would get the validation for unique feature names in the FE but not on the Ludwig side if we removed this check.

File: ludwig/config_validation/checks.py
Comment (ksbrar) [reply ludwig/config_validation/checks.py:206]: +1

File: ludwig/config_validation/checks.py
Comment (ksbrar) [reply ludwig/config_validation/checks.py:214]: +1

File: ludwig/config_validation/checks.py
Comment (tgaddair) [reply ludwig/config_validation/checks.py:155]: Yep, good call. I think it's fine to keep this in, then.

File: ludwig/config_validation/checks.py
Comment (tgaddair) [inline ludwig/config_validation/checks.py:322]: All these split checks are run after we have generated the `ModelConfig` object, but for some reason we're running them against untyped dicts instead of the config object? We should change these to operate on the `ModelConfig` object for better type safety.

File: ludwig/config_validation/validation.py
Comment (tgaddair) [inline ludwig/config_validation/validation.py:98]: As described above, should change this to:

```
get_config_check_registry().check_config(model_config)
```

So validate on the config object, not on a dict.

Furthermore, since this is useful to check anytime we're creating a `ModelConfig`, this should really be run as part of the `ModelConfig.from_dict()` function at the end. That way, any user that wants to create a `ModelConfig` will benefit from these checks.

File: ludwig/error.py
Comment (tgaddair) [inline ludwig/error.py:53]: You need to make this error type serializable. See https://github.com/ludwig-ai/ludwig/pull/2695.

In doing so, please add a unit test similar to the one from the linked PR to verify that serialization works.

File: tests/integration_tests/test_gbm.py
Comment (tgaddair) [inline tests/integration_tests/test_gbm.py:87]: Are these tests all replicated elsewhere?

**Review by tgaddair — CHANGES_REQUESTED** [review]
Changes requested:

- Use `ModelConfig` in these validation hooks, not a dict
- Run validation hooks at the end of `ModelConfig.from_dict()` instead of within the standalone validation function
- Make `ConfigValidationError` serializable and add a unit test
- Wrap mashmallow ValidationErrors with `ConfigValidationError` to make the exception type consistent.

File: tests/ludwig/config_validation/test_validate_config_combiner.py
Comment (tgaddair) [inline tests/ludwig/config_validation/test_validate_config_combiner.py:115]: Why drop the match? Is there a way to get this back in? Otherwise, we lose confidence that we're actually testing the right thing.

**Review by tgaddair — COMMENTED** [review]
As discussed, would still be cool to move this into the ModelConfig, otherwise it's not teribly useful.

File: tests/ludwig/config_validation/test_validate_config_combiner.py
Comment (justinxzhao) [reply tests/ludwig/config_validation/test_validate_config_combiner.py:115]: I removed the error strings at first because some of them had changed slightly while adding aux checks, and I wanted to minimize noise during development.

After removing the regex strings, however, the tests felt easier to maintain and cleaner overall.

Testing for specific error string regex'es gives us some additional fidelity, but on the whole to me it feels like testing an implementation detail, where the maintenance cost might not outweigh the increased specificity. Happy to add these matches back though if you feel strongly about it.

**Comment by justinxzhao** [comment]
> As discussed, would still be cool to move this into the ModelConfig, otherwise it's not teribly useful.

I'd prefer to make that change in a separate PR if that's ok. I anticipate that folding schema validation and aux checks into the model config will affect several more tests, as every config in ludwig tests will be validated, which we've never tested explicitly for before.

**Comment by abidwael** [comment]
@justinxzhao you can merge with latest master to fix https://github.com/ludwig-ai/ludwig/actions/runs/4178266592/jobs/7236853564#step:12:4290

File: tests/ludwig/config_validation/test_validate_config_features.py
Comment (tgaddair) [inline tests/ludwig/config_validation/test_validate_config_features.py:141]: Comment is incorrect.

File: tests/ludwig/utils/test_errors.py
Comment (tgaddair) [inline tests/ludwig/utils/test_errors.py:22]: Remove print.

**Review by tgaddair — APPROVED** [review]
LGTM, please address nits before landing.

File: tests/ludwig/config_validation/test_validate_config_features.py
Comment (justinxzhao) [reply tests/ludwig/config_validation/test_validate_config_features.py:141]: Done.

File: tests/ludwig/utils/test_errors.py
Comment (justinxzhao) [reply tests/ludwig/utils/test_errors.py:22]: Done.

**Pull request merged**
